pinot-core/src/main/java/com/linkedin/pinot/core/minion/BackfillDateTimeColumn.java

```diff
@@ -0,0 +1,263 @@
+  private final DateTimeFieldSpec _destDateTimeFieldSpec;
+
+  public BackfillDateTimeColumn(@Nonnull File originalIndexDir, @Nonnull File backfilledIndexDir,
+      @Nonnull TimeFieldSpec srcTimeSpec, @Nonnull DateTimeFieldSpec destDateTimeSpec)
+      throws Exception {
+    _originalIndexDir = originalIndexDir;
+    _backfilledIndexDir = backfilledIndexDir;
+    _srcTimeFieldSpec = srcTimeSpec;
+    _destDateTimeFieldSpec = destDateTimeSpec;
+  }
+
+  public boolean backfill() throws Exception {
+
+    boolean success = true;
+    SegmentMetadataImpl originalSegmentMetadata = new SegmentMetadataImpl(_originalIndexDir);
+    String segmentName = originalSegmentMetadata.getName();
+    String tableName = originalSegmentMetadata.getTableName();
+    LOGGER.info("Start backfilling segment: {} in table: {}", segmentName, tableName);
+
+    PinotSegmentRecordReader segmentRecordReader = new PinotSegmentRecordReader(_originalIndexDir);
+    // read only rawdocs, everything after that, if present, is generated by star tree
+    BackfillDateTimeRecordReader wrapperReader =
+        new BackfillDateTimeRecordReader(segmentRecordReader, _srcTimeFieldSpec,
+            _destDateTimeFieldSpec, originalSegmentMetadata.getTotalRawDocs());
+    LOGGER.info("Segment dir: {} Output Dir: {}", _originalIndexDir.getAbsolutePath(),
+        _backfilledIndexDir.getAbsolutePath());
+
+    LOGGER.info("Creating segment generator config for {}", segmentName);
+    SegmentGeneratorConfig config = new SegmentGeneratorConfig();
+    config.setInputFilePath(_originalIndexDir.getAbsolutePath());
+    config.setFormat(FileFormat.PINOT);
+    config.setOutDir(_backfilledIndexDir.getAbsolutePath());
+    config.setOverwrite(true);
+    config.setTableName(tableName);
+    config.setSegmentName(segmentName);
+    config.setSchema(wrapperReader.getSchema());
+
+    StarTreeMetadata starTreeMetadata = originalSegmentMetadata.getStarTreeMetadata();
+    if (starTreeMetadata != null) {
+      config.setEnableStarTreeIndex(true);
+      StarTreeIndexSpec starTreeIndexSpec = new StarTreeIndexSpec();
+      starTreeIndexSpec.setDimensionsSplitOrder(starTreeMetadata.getDimensionsSplitOrder());
+      starTreeIndexSpec.setMaxLeafRecords((int) starTreeMetadata.getMaxLeafRecords());
+      starTreeIndexSpec.setSkipMaterializationCardinalityThreshold((int) starTreeMetadata
+          .getSkipMaterializationCardinality());
+      starTreeIndexSpec.setSkipMaterializationForDimensions(Sets.newHashSet(starTreeMetadata
+          .getSkipMaterializationForDimensions()));
+      starTreeIndexSpec.setSkipStarNodeCreationForDimensions(Sets.newHashSet(starTreeMetadata
+          .getSkipStarNodeCreationForDimensions()));
+      config.setStarTreeIndexSpec(starTreeIndexSpec);
+    }
+
+    LOGGER.info("Creating segment for {} with config {}", segmentName, config.toString());
+    SegmentIndexCreationDriverImpl driver = new SegmentIndexCreationDriverImpl();
```

Comment:
It may be useful to set up SegmentIndexCreationDriverImpl (via config) so that it can log a message every (say N) rows. I understand this is not something you did, but maybe it is useful to add that information. You decide.